tests/constraints/tests.py

```diff
@@ -365,6 +366,34 @@ def test_validate_pk_field(self):
             constraint_with_pk.validate(ChildModel, ChildModel(id=1, age=1))
         constraint_with_pk.validate(ChildModel, ChildModel(pk=1, age=1), exclude={"pk"})
 
+    @skipUnlessDBFeature("supports_table_check_constraints", "supports_json_field")
+    def test_database_constraint_json(self):
+        JSONConstraintModel.objects.create(
+            data={"release": "5.0.2", "version": "stable", "product_name": "django"}
+        )
+        with self.assertRaises(IntegrityError):
+            JSONConstraintModel.objects.create(
+                data={
+                    "release": "5.0.2",
+                    "version": "not stable",
+                    "product_name": "django",
+                }
+            )
+
+    @skipUnlessDBFeature("supports_json_field")
+    def test_validate_json_exact_constraint(self):
+        data = {"release": "5.0.2", "version": "stable", "product_name": "django"}
+        json_exact_constraint = models.CheckConstraint(
+            check=models.Q(data__version="stable"),
+            name="only_stable_version",
+        )
+        json_exact_constraint.validate(JSONFieldModel, JSONFieldModel(data=data))
+
+        data = {"release": "5.0.2", "version": "not stable", "product_name": "django"}
+        msg = f"Constraint “{json_exact_constraint.name}” is violated."
+        with self.assertRaisesMessage(ValidationError, msg):
+            json_exact_constraint.validate(JSONFieldModel, JSONFieldModel(data=data))
```

Comment:
I'd move these tests to the `tests/schema` since it's mainly about crash when adding a constraint on Oracle.

Reply:
For example:
```diff
diff --git a/tests/schema/tests.py b/tests/schema/tests.py
index a16c4a6bfe..70613f0b45 100644
--- a/tests/schema/tests.py
+++ b/tests/schema/tests.py
@@ -2755,6 +2755,40 @@ class SchemaTests(TransactionTestCase):
             DurationModel.objects.create(duration=datetime.timedelta(minutes=4))
         DurationModel.objects.create(duration=datetime.timedelta(minutes=10))
 
+    @skipUnlessDBFeature(
+        "supports_column_check_constraints",
+        "can_introspect_check_constraints",
+        "supports_json_field",
+    )
+    @isolate_apps("schema")
+    def test_check_constraint_exact_jsonfield(self):
+        class JSONConstraintModel(Model):
+            data = JSONField()
+
+            class Meta:
+                app_label = "schema"
+
+        with connection.schema_editor() as editor:
+            editor.create_model(JSONConstraintModel)
+        self.isolated_local_models = [JSONConstraintModel]
+        constraint_name = "check_only_stable_version"
+        constraint = CheckConstraint(
+            check=Q(data__version="stable"),
+            name=constraint_name,
+        )
+        JSONConstraintModel._meta.constraints = [constraint]
+        with connection.schema_editor() as editor:
+            editor.add_constraint(JSONConstraintModel, constraint)
+        constraints = self.get_constraints(JSONConstraintModel._meta.db_table)
+        self.assertIn(constraint_name, constraints)
+        with self.assertRaises(IntegrityError), atomic():
+            JSONConstraintModel.objects.create(
+                data={"release": "5.0.2dev", "version": "dev"}
+            )
+        JSONConstraintModel.objects.create(
+            data={"release": "5.0.3", "version": "stable"}
+        )
+
     @skipUnlessDBFeature(
         "supports_column_check_constraints", "can_introspect_check_constraints"
     )

```
:point_up: This also won't fail on PostgreSQL and allow us to handle it separately.

Reply:
I have applied your patch but wasn't sure what to do with `test_validate_json_exact_constraint`, so I left it in `tests/constraints/tests.py` because there are also other `constraint.validate` tests for JSONField.

Reply:
I'd like to discuss `psycopg2` failure separately.